### ISSUE TITLE
DNS lookup failure for: host.docker.internal

### ISSUE THREAD

**Issue by MinkiPan** [docker/for-mac#3742]
> The proxy server received an invalid response from an upstream server.
> The proxy server could not handle the request GET /.
> 
> Reason: DNS lookup failure for: host.docker.internal

I think this is similar to this one #2965

I urgently need help...

macOS Mojave
10.14.5

Docker Desktop 2.0.0.3
Engine: 18.09.2

**Comment by tjdavey** [comment]
Does `docker.for.mac.localhost` resolve for you? My experience with this continuing issue is that `docker.for.mac.localhost` continues to work when `host.docker.internal` does not.

**Comment by tdtm** [comment]
Neither works for me.

Engine 19.03.1, build 74b1e89
Machine 0.16.1
Desktop 2.1.0.1

On MacOS 10.14.6 (18G87)

**Comment by rossjp** [comment]
Neither works for me either:
```
$ docker run --rm alpine nslookup host.docker.internal
nslookup: can't resolve '(null)': Name does not resolve
nslookup: can't resolve 'host.docker.internal': Name does not resolve

$ docker run --rm alpine nslookup docker.for.mac.localhost

nslookup: can't resolve '(null)': Name does not resolve
nslookup: can't resolve 'docker.for.mac.localhost': Name does not resolve
```

- Docker version 19.03.2, build 6a30dfc
- docker-machine version 0.16.2, build bd45ab13
- Docker Desktop: Version 2.1.0.3 (38240), Channel stable
- MacOS 10.14.5

Perhaps of interest: `host.docker.internal` **was** working for me before I performed a MacOS restart. Now it does not.

**Comment by simonlin527** [comment]
I have the same issue on 
* MacOS 10.14.6
* Docker version 19.03.2 and 18.09.2, build 6247962. 
* Docker Desktop: Version 2.0.0.3 (31259)
* Docker machine: 0.16.1

My current workaround is to hard code wherever I need to use the host ip. Horrible solution, but it works for now.

**Comment by jmmen1** [comment]
I've had the same issue on both macos 10.14 and 10.15

Current setup is:
- macos 10.15.1
- Docker version 19.03.5, build 633a0ea
- Docker Desktop: Version 2.1.0.5 (40693)
- Docker Machine: 0.16.2

I've noticed that a clean install of Docker Desktop will resolve host.docker.internal. But, if I specify my company's DNS (or any DNS, really) via the docker run --dns option or by adding DNS entries in daemon.json, then host.docker.internal no longer resolves.

Also, it seems that docker.for.mac.localhost resolves to 127.0.0.1 inside the container, not to the Mac host IP.

**Comment by mattca** [comment]
I experienced the same issue as @jmmen1 , same versions and the same thing where the problem would only occur when adding DNS entries.

I noticed that `nameserver 192.168.65.1` was getting removed from the `/etc/resolv.conf` file in the started container when the custom DNS entries were present.

So if I added "192.168.65.1" as an extra DNS entry to the daemon.json with the other ones I wanted to define, then `host.docker.internal` would work as expected.

It might be that that IP is a default DNS entry, and when you define custom DNS entries via daemon.json, then the custom ones override that and it gets removed.

Can't confirm how reliable this is.
